rpc/src/rpc_subscriptions.rs

```diff
@@ -785,100 +857,204 @@ impl RpcSubscriptions {
-                        commitment_slots,
-                        |bank, params| bank.get_account_modified_slot(&params.pubkey),
-                        filter_account_result,
-                        notifier,
-                        false,
-                    );
-
                     num_accounts_found.fetch_add(1, Ordering::Relaxed);
+                    if let Some(slot) = slot {
+                        let notified = check_commitment_and_notify(
+                            params,
+                            subscription,
+                            bank_forks,
+                            slot,
+                            |bank, params| bank.get_account_modified_slot(&params.pubkey),
+                            filter_account_result,
+                            notifier,
+                            false,
+                        );
+
+                        if notified {
+                            num_accounts_notified.fetch_add(1, Ordering::Relaxed);
+                        }
+                    }
+                }
+                SubscriptionParams::Block(params) => {
+                    num_blocks_found.fetch_add(1, Ordering::Relaxed);
+                    if let Some(slot) = slot {
+                        // We're calling it unnotified in this context
+                        // because, logically, it gets set to `last_notified_slot + 1`
+                        // on the final iteration of the loop down below.
+                        // This is used to notify blocks for slots that were
+                        // potentially missed due to upstream transient errors
+                        // that led to this notification not being triggered for
+                        // a slot.
+                        //
+                        // e.g.
+                        // notify_watchers is triggered for Slot 1
+                        // some time passes
+                        // notify_watchers is triggered for Slot 4
+                        // this will try to fetch blocks for slots 2, 3, and 4
+                        // as long as they are ancestors of `slot`
+                        let mut w_last_unnotified_slot =
+                            subscription.last_notified_slot.write().unwrap();
+                        // would mean it's the first notification for this subscription connection
+                        if *w_last_unnotified_slot == 0 {
+                            *w_last_unnotified_slot = slot;
+                        }
 
-                    if notified {
-                        num_accounts_notified.fetch_add(1, Ordering::Relaxed);
+                        let ancestors = bank_forks.read().unwrap().ancestors();
+                        let ancestors = ancestors.get(&slot);
+                        let empty_set = &HashSet::new();
+                        let ancestors = ancestors.unwrap_or(empty_set);
+                        for s in *w_last_unnotified_slot..slot + 1 {
```

Comment:
So I was thinking of going this route but opted out just because of time complexity. It may be negligible (maybe not?) But from a glance this would be O(n*log(n)) cause of the sort and still iterates through all ancestors in filter vs. O(1) access of the HashSet approach. Perhaps the `sort` can be omitted completely if we do something like:

```
    let mut slots_to_notify: Vec<_> = (w_last_unnotified_slot..slot).collect();
    slots_to_notify = slots_to_notify.into_iter().filter(|slot| ancestors.contains(slot)).collect();
    slots_to_notify.push(slot);
```
Wdyt of the three approaches?

Reply:
(Sorry for delay!) I think it may be negligible, but fair point. I like approach 3 without the sort the best!

Reply:
actually I have an idea!